### ISSUE TITLE
Can not change max_input_len of Encoder while building engine in Encoder-Decoder model (T5)

### ISSUE THREAD

**Issue by thanhlt998** [NVIDIA/TensorRT-LLM#1617]
I built the engines for T5 model with the following scripts for the latest version of TensorRT-LLM:

```
export MODEL_DIR="path_to_t5_model" # or "flan-t5-small"
export MODEL_NAME="t5model"
export MODEL_TYPE="t5"
export INFERENCE_PRECISION="float16"
export TP_SIZE=1
export PP_SIZE=1
export WORLD_SIZE=1
export MAX_BEAM_WIDTH=1
export OUTPUT_DIR="triton_model_repos/${MODEL_NAME}/tensorrt_llm/1"

trtllm-build --checkpoint_dir "${MODEL_DIR}/trt_models/${INFERENCE_PRECISION}/tp${TP_SIZE}/pp${PP_SIZE}/encoder" \
                --output_dir "${OUTPUT_DIR}/${WORLD_SIZE}-gpu/${INFERENCE_PRECISION}/tp${TP_SIZE}/encoder" \
                --paged_kv_cache enable \
                --moe_plugin disable \
                --enable_xqa disable \
                --use_custom_all_reduce disable \
                --max_beam_width ${MAX_BEAM_WIDTH} \
                --max_batch_size 8 \
                --max_input_len 4096 \
                --max_encoder_input_len 4096 \
                --max_output_len 4096 \
                --gemm_plugin ${INFERENCE_PRECISION} \
                --bert_attention_plugin ${INFERENCE_PRECISION} \
                --gpt_attention_plugin ${INFERENCE_PRECISION} \
                --remove_input_padding disable \
                --context_fmha disable

# For decoder, refer to the above content and set --max_input_len correctly
trtllm-build --checkpoint_dir "${MODEL_DIR}/trt_models/${INFERENCE_PRECISION}/tp${TP_SIZE}/pp${PP_SIZE}/decoder" \
                --output_dir "${OUTPUT_DIR}/${WORLD_SIZE}-gpu/${INFERENCE_PRECISION}/tp${TP_SIZE}/decoder" \
                --paged_kv_cache enable \
                --moe_plugin disable \
                --enable_xqa disable \
                --use_custom_all_reduce disable \
                --max_beam_width ${MAX_BEAM_WIDTH} \
                --max_batch_size 8 \
                --max_output_len 4096 \
                --max_encoder_input_len 4096 \
                --gemm_plugin ${INFERENCE_PRECISION} \
                --bert_attention_plugin ${INFERENCE_PRECISION} \
                --gpt_attention_plugin ${INFERENCE_PRECISION} \
                --remove_input_padding disable \
                --context_fmha disable \
                --max_input_len 1
```

Then I run inference with the built engines, it only works with input with the length <= 1024 although I built with `--max_input_len=4096` and `max_encoder_input_len=4096`. When I run with the inputs (length > 1024), it raises the following error in the triton server log:

```
[05/16/2024-08:43:31] [TRT] [E] 3: [executionContext.cpp::setInputShape::2068] Error Code 3: API Usage Error (Parameter check failed at: runtime/api/executionContext.cpp::setInputShape::2068, condition: satisfyProfile Runtime dimension does not satisfy any optimization profile.)
``` 

I also edit some bug in line https://github.com/NVIDIA/TensorRT-LLM/blob/bf0a5afc92f4b2b3191e9e55073953c1f600cf2d/tensorrt_llm/runtime/session.py#L170 (`f"engine supports [min, opt, max] = {self.engine.get_profile_shape(context.active_optimization_profile, name)}"` -> `f"engine supports [min, opt, max] = {self.engine.get_tensor_profile_shape(name, context.active_optimization_profile)}"`) to show the error:

```
I0516 07:36:03.073916 35685 pb_stub.cc:715] Failed to process the request(s) for model 'tensorrt_llm_0_0', message: ValueError: Couldn't assign input_ids with shape torch.Size([1, 1398]), engine supports [min, opt, max] = [(1, 1), (4, 512), (8, 1024)]
```

It seems like max_input_len 1024 set somewhere in the code?

**Comment by jhaotingc** [comment]
Hi @thanhlt998,

Would you mind sharing the commit you are trying on TensorRT-LLM, as well as the hardware specifications you are using?
I'm trying to reproduce your error but did not see this in the latest commit: `5d8ca2faf74c494f220c8f71130340b513eea9a9`

Steps:
1. Follow the installation steps on [building from source](https://nvidia.github.io/TensorRT-LLM/installation/build-from-source-linux.html#build-tensorrt-llm).
```
git clone https://github.com/NVIDIA/TensorRT-LLM.git
cd TensorRT-LLM
git submodule update --init --recursive
git lfs pull
docker build --pull  \
            --target devel \
            --file docker/Dockerfile.multi \
            --tag tensorrt_llm/devel:latest \
            .
docker run --rm -it \
        --ipc=host --ulimit memlock=-1 --ulimit stack=67108864 --gpus=all \
        --volume ${PWD}:/code/tensorrt_llm \
        --workdir /code/tensorrt_llm \
        tensorrt_llm/devel:latest
python3 ./scripts/build_wheel.py --trt_root /usr/local/tensorrt
pip install ./build/tensorrt_llm*.whl
```
2. Build the model as your script above. You'll notice an error which I just notice after trying your script.
```
export MODEL_NAME="t5-small"
export MODEL_TYPE="t5"
export INFERENCE_PRECISION="float16"
export TP_SIZE=1
export PP_SIZE=1
export WORLD_SIZE=1
export MAX_BEAM_WIDTH=1

python convert_checkpoint.py --model_type ${MODEL_TYPE} \
                --model_dir tmp/hf_models/${MODEL_NAME} \
                --output_dir tmp/trt_models/${MODEL_NAME}/${INFERENCE_PRECISION} \
                --tp_size ${TP_SIZE} \
                --pp_size ${PP_SIZE} \
                --weight_data_type float32 \
                --dtype ${INFERENCE_PRECISION}

trtllm-build --checkpoint_dir tmp/trt_models/${MODEL_NAME}/${INFERENCE_PRECISION}/tp${TP_SIZE}/pp${PP_SIZE}/encoder \
                --output_dir tmp/trt_engines/${MODEL_NAME}/${WORLD_SIZE}-gpu/${INFERENCE_PRECISION}/tp${TP_SIZE}/encoder \
                --paged_kv_cache enable \
                --moe_plugin disable \
                --enable_xqa disable \
                --use_custom_all_reduce disable \
                --max_beam_width ${MAX_BEAM_WIDTH} \
                --max_input_len 4096 \
                --max_batch_size 8 \
                --max_output_len 4096 \
                --gemm_plugin ${INFERENCE_PRECISION} \
                --bert_attention_plugin ${INFERENCE_PRECISION} \
                --gpt_attention_plugin ${INFERENCE_PRECISION} \
                --remove_input_padding disable \
                --context_fmha disable

trtllm-build --checkpoint_dir tmp/trt_models/${MODEL_NAME}/${INFERENCE_PRECISION}/tp${TP_SIZE}/pp${PP_SIZE}/decoder \
                --output_dir tmp/trt_engines/${MODEL_NAME}/${WORLD_SIZE}-gpu/${INFERENCE_PRECISION}/tp${TP_SIZE}/decoder \
                --paged_kv_cache enable \
                --moe_plugin disable \
                --enable_xqa disable \
                --use_custom_all_reduce disable \
                --max_beam_width ${MAX_BEAM_WIDTH} \
                --max_batch_size 8 \
                --max_output_len 4096 \
                --max_encoder_input_len 4096 \
                --gemm_plugin ${INFERENCE_PRECISION} \
                --bert_attention_plugin ${INFERENCE_PRECISION} \
                --gpt_attention_plugin ${INFERENCE_PRECISION} \
                --remove_input_padding disable \
                --context_fmha disable \
                --max_input_len 1

```

You'll see an error when building decoder, as listed here: `UnboundLocalError: local variable 'kv' referenced before assignment`. This triggers when `--remove_input_padding disable` and `--paged_kv_cache enable` are set together. I'm working on a quick fix right now, if you're in a hurry, you can move [tensorrt_llm/models/enc_dec/model.py#L1666~L1670](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/models/enc_dec/model.py#L1666~L1670) to [L1584](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/models/enc_dec/model.py#L1584), then replace the file to the one binded to tensorrt_llm package like so:
```
cp tensorrt_llm/models/enc_dec/model.py /home/dev_user/.local/lib/python3.10/site-packages/tensorrt_llm/models/enc_dec/
```

Re-run the script and you should get the model that can take 4096 input.

3. Use your own input in [run.py](https://github.com/NVIDIA/TensorRT-LLM/blob/main/examples/enc_dec/run.py#L692). I'm seeing the engine behave as intended. Ran with no problem when `input lengths:  tensor([4095])` but failed with `> 4096` with the same error you find, which is intended.


Other than that, I'm suspecting that you're using tritonserver, which may not be the latest version of TensorRT-LLM. You can find the version when you run any TRTLLM related script as shown below. It'll be very helpful for me to investigate.
```
[TensorRT-LLM] TensorRT-LLM version: 0.11.0.dev2024052100                                                                                                                                                                                          
```

Thanks!!.

**Comment by thanhlt998** [comment]
Thank @jtchen0528 for your response!
I am using TensorRT-LLM version 0.10.0.dev2024050700. I rebuilt triton backend with TensorRT-LLM in commit e239adcfbf8e72bf1eefaa166781343f41770d70. I found the bug in both RTX2080Ti and RTX4090 hardware specifications. 
I have already commented the line 1666-1670 in order to build the engine successfully.

**Comment by jhaotingc** [comment]
Hi @thanhlt998,

I tracked down [tensorrtllm_backend](https://github.com/triton-inference-server/tensorrtllm_backend.git) at commit `e239adcfbf8e72bf1eefaa166781343f41770d70`, which has the submodule [TensorRT-LLM](https://github.com/NVIDIA/TensorRT-LLM/tree/89ba1b1a67d570e41b03da87e5518eaff0d31fbf) commit `89ba1b1a67d570e41b03da87e5518eaff0d31fbf`. I then built the container with [building from source](https://nvidia.github.io/TensorRT-LLM/installation/build-from-source-linux.html#build-tensorrt-llm) on an RTX 4090, built the engine with your script. Ran the engine with a length 4095 input. I still did not see the error (other then the line 1666-1670 bug, which will be merged in a week).

I'm now suspecting your engine does not get copied to the correct folder. Triton server takes a specific engine folder listed in [this file](https://github.com/triton-inference-server/tensorrtllm_backend/blob/e239adcfbf8e72bf1eefaa166781343f41770d70/all_models/inflight_batcher_llm/tensorrt_llm/config.pbtxt#L366) in your triton server config.

**Comment by thanhlt998** [comment]
Hi @jtchen0528 ,
I have just tried again with t5-small, the built engine ran successfully with a length 4095. The T5 model I tried has ~1.5B parameters, it is larger than t5-small. So I tried to re-build the engine with `--max_batch_size 2 --max_input_len 4096 --max_encoder_input_len 4096 --max_output_len 4096` instead of `--max_batch_size 8`, I can run the engine with a length 4095 input.
But when I built the engine with `--max_batch_size 8`, it was successful without any errors. It only raises the error `engine supports [min, opt, max] = [(1, 1), (4, 512), (8, 1024)]` when run the built engine with the input_length > 1024.
The above experiment I ran with 2080Ti GPU hardware and the above source code commits.

So it need to raise error in the building step instead of running step???

**Comment by jhaotingc** [comment]
Would you mind sharing which T5 model you're running, and the steps to run the built engine so that I can see your error? 
As I shared in the comment, I can build a 8 batch engine and run without problem.

> Hi @thanhlt998,
> 
> Would you mind sharing the commit you are trying on TensorRT-LLM, as well as the hardware specifications you are using? I'm trying to reproduce your error but did not see this in the latest commit: `5d8ca2faf74c494f220c8f71130340b513eea9a9`
> 
> Steps:
> 
> 1. Follow the installation steps on [building from source](https://nvidia.github.io/TensorRT-LLM/installation/build-from-source-linux.html#build-tensorrt-llm).
> 
> ```
> git clone https://github.com/NVIDIA/TensorRT-LLM.git
> cd TensorRT-LLM
> git submodule update --init --recursive
> git lfs pull
> docker build --pull  \
>             --target devel \
>             --file docker/Dockerfile.multi \
>             --tag tensorrt_llm/devel:latest \
>             .
> docker run --rm -it \
>         --ipc=host --ulimit memlock=-1 --ulimit stack=67108864 --gpus=all \
>         --volume ${PWD}:/code/tensorrt_llm \
>         --workdir /code/tensorrt_llm \
>         tensorrt_llm/devel:latest
> python3 ./scripts/build_wheel.py --trt_root /usr/local/tensorrt
> pip install ./build/tensorrt_llm*.whl
> ```
> 
> 2. Build the model as your script above. You'll notice an error which I just notice after trying your script.
> 
> ```
> export MODEL_NAME="t5-small"
> export MODEL_TYPE="t5"
> export INFERENCE_PRECISION="float16"
> export TP_SIZE=1
> export PP_SIZE=1
> export WORLD_SIZE=1
> export MAX_BEAM_WIDTH=1
> 
> python convert_checkpoint.py --model_type ${MODEL_TYPE} \
>                 --model_dir tmp/hf_models/${MODEL_NAME} \
>                 --output_dir tmp/trt_models/${MODEL_NAME}/${INFERENCE_PRECISION} \
>                 --tp_size ${TP_SIZE} \
>                 --pp_size ${PP_SIZE} \
>                 --weight_data_type float32 \
>                 --dtype ${INFERENCE_PRECISION}
> 
> trtllm-build --checkpoint_dir tmp/trt_models/${MODEL_NAME}/${INFERENCE_PRECISION}/tp${TP_SIZE}/pp${PP_SIZE}/encoder \
>                 --output_dir tmp/trt_engines/${MODEL_NAME}/${WORLD_SIZE}-gpu/${INFERENCE_PRECISION}/tp${TP_SIZE}/encoder \
>                 --paged_kv_cache enable \
>                 --moe_plugin disable \
>                 --enable_xqa disable \
>                 --use_custom_all_reduce disable \
>                 --max_beam_width ${MAX_BEAM_WIDTH} \
>                 --max_input_len 4096 \
>                 --max_batch_size 8 \
>                 --max_output_len 4096 \
>                 --gemm_plugin ${INFERENCE_PRECISION} \
>                 --bert_attention_plugin ${INFERENCE_PRECISION} \
>                 --gpt_attention_plugin ${INFERENCE_PRECISION} \
>                 --remove_input_padding disable \
>                 --context_fmha disable
> 
> trtllm-build --checkpoint_dir tmp/trt_models/${MODEL_NAME}/${INFERENCE_PRECISION}/tp${TP_SIZE}/pp${PP_SIZE}/decoder \
>                 --output_dir tmp/trt_engines/${MODEL_NAME}/${WORLD_SIZE}-gpu/${INFERENCE_PRECISION}/tp${TP_SIZE}/decoder \
>                 --paged_kv_cache enable \
>                 --moe_plugin disable \
>                 --enable_xqa disable \
>                 --use_custom_all_reduce disable \
>                 --max_beam_width ${MAX_BEAM_WIDTH} \
>                 --max_batch_size 8 \
>                 --max_output_len 4096 \
>                 --max_encoder_input_len 4096 \
>                 --gemm_plugin ${INFERENCE_PRECISION} \
>                 --bert_attention_plugin ${INFERENCE_PRECISION} \
>                 --gpt_attention_plugin ${INFERENCE_PRECISION} \
>                 --remove_input_padding disable \
>                 --context_fmha disable \
>                 --max_input_len 1
> ```
> 
> You'll see an error when building decoder, as listed here: `UnboundLocalError: local variable 'kv' referenced before assignment`. This triggers when `--remove_input_padding disable` and `--paged_kv_cache enable` are set together. I'm working on a quick fix right now, if you're in a hurry, you can move [tensorrt_llm/models/enc_dec/model.py#L1666~L1670](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/models/enc_dec/model.py#L1666~L1670) to [L1584](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/models/enc_dec/model.py#L1584), then replace the file to the one binded to tensorrt_llm package like so:
> 
> ```
> cp tensorrt_llm/models/enc_dec/model.py /home/dev_user/.local/lib/python3.10/site-packages/tensorrt_llm/models/enc_dec/
> ```
> 
> Re-run the script and you should get the model that can take 4096 input.
> 
> 3. Use your own input in [run.py](https://github.com/NVIDIA/TensorRT-LLM/blob/main/examples/enc_dec/run.py#L692). I'm seeing the engine behave as intended. Ran with no problem when `input lengths:  tensor([4095])` but failed with `> 4096` with the same error you find, which is intended.
> 
> Other than that, I'm suspecting that you're using tritonserver, which may not be the latest version of TensorRT-LLM. You can find the version when you run any TRTLLM related script as shown below. It'll be very helpful for me to investigate.
> 
> ```
> [TensorRT-LLM] TensorRT-LLM version: 0.11.0.dev2024052100                                                                                                                                                                                          
> ```
> 
> Thanks!!.

**Comment by thanhlt998** [comment]
@jtchen0528 , I run my pretrained T5 model with the following config:
```
{
  "architectures": [
    "T5ForConditionalGeneration"
  ],
  "classifier_dropout": 0.0,
  "d_ff": 3840,
  "d_kv": 64,
  "d_model": 1536,
  "decoder_start_token_id": 0,
  "dense_act_fn": "silu",
  "dropout_rate": 0.1,
  "eos_token_id": 1,
  "extra_ids": 256,
  "feed_forward_proj": "gated-silu",
  "initializer_factor": 1.0,
  "is_encoder_decoder": true,
  "is_gated_act": true,
  "layer_norm_epsilon": 1e-06,
  "model_type": "t5",
  "n_positions": 2048,
  "num_decoder_layers": 24,
  "num_heads": 24,
  "num_layers": 24,
  "output_past": true,
  "pad_token_id": 0,
  "relative_attention_max_distance": 128,
  "relative_attention_num_buckets": 32,
  "tie_word_embeddings": false,
  "torch_dtype": "float32",
  "transformers_version": "4.39.1",
  "use_cache": true,
  "vocab_size": 65664
}
```
To see the errors I just follow `run.py` of the enc-dec example code and increase the length of the samples to have input length > 1024

**Comment by jhaotingc** [comment]
@thanhlt998 Is the config same as any T5 family on HuggingFace?

**Comment by thanhlt998** [comment]
@jtchen0528 I customized the config based on my demand, you can try the google/flan-t5-xl (~3B parameters) to reproduce the errors.

**Comment by jhaotingc** [comment]
@thanhlt998 I need more detailed reproduction steps. I just tried with flan-t5-xl without problem. I ran on H100 for this because a 4090 cannot hold a flan-t5-xl but I don't think the hardware matters with this bug.
The exact same script is provided above, just change the model to google/flan-t5-xl on huggingface.
I'm also attaching the script I run here. [test.txt](https://github.com/NVIDIA/TensorRT-LLM/files/15428807/test.txt)
